lib/experimental/class-theme-json-webfonts-handler.php

```diff
@@ -0,0 +1,537 @@
+				if ( ! in_array( $webfont['font-display'], array( 'auto', 'block', 'fallback', 'swap' ), true ) ) {
+					$webfont['font-display'] = 'fallback';
+				}
+
+				$valid_props = array(
+					'ascend-override',
+					'descend-override',
+					'font-display',
+					'font-family',
+					'font-stretch',
+					'font-style',
+					'font-weight',
+					'font-variant',
+					'font-feature-settings',
+					'font-variation-settings',
+					'line-gap-override',
+					'size-adjust',
+					'src',
+					'unicode-range',
+				);
+
+				foreach ( $webfont as $prop => $value ) {
+					if ( ! in_array( $prop, $valid_props, true ) ) {
+						unset( $webfont[ $prop ] );
+					}
+				}
+
+				return $webfont;
+			}
+
+			/**
+			 * Generate and enqueue webfonts styles.
+			 *
+			 * @since 6.0.0
+			 */
+			public function generate_and_enqueue_styles() {
+				// Generate the styles.
+				$styles = $this->get_css();
+
+				// Bail out if there are no styles to enqueue.
+				if ( '' === $styles ) {
+					return;
+				}
+
+				// Enqueue the stylesheet.
+				wp_register_style( $this->stylesheet_handle, '' );
+				wp_enqueue_style( $this->stylesheet_handle );
+
+				// Add the styles to the stylesheet.
+				wp_add_inline_style( $this->stylesheet_handle, $styles );
+			}
+
+			/**
+			 * Generate and enqueue editor styles.
+			 *
+			 * @since 6.0.0
+			 */
+			public function generate_and_enqueue_editor_styles() {
+				// Generate the styles.
+				$styles = $this->get_css();
+
+				// Bail out if there are no styles to enqueue.
+				if ( '' === $styles ) {
+					return;
+				}
+
+				wp_add_inline_style( 'wp-block-library', $styles );
+			}
+
+			/**
+			 * Gets the `@font-face` CSS styles for locally-hosted font files.
+			 *
+			 * @since 6.0.0
+			 *
+			 * @return string The `@font-face` CSS.
+			 */
+			public function get_css() {
+				$css = '';
+
+				foreach ( $this->webfonts as $webfont ) {
+					// Order the webfont's `src` items to optimize for browser support.
+					$webfont = $this->order_src( $webfont );
+
+					// Build the @font-face CSS for this webfont.
+					$css .= '@font-face{' . $this->build_font_face_css( $webfont ) . '}';
+				}
+
+				return $css;
+			}
+
+			/**
+			 * Order `src` items to optimize for browser support.
+			 *
+			 * @since 6.0.0
+			 *
+			 * @param array $webfont Webfont to process.
+			 *
+			 * @return array Ordered `src` items.
+			 */
+			private function order_src( array $webfont ) {
+				$src         = array();
+				$src_ordered = array();
+
+				foreach ( $webfont['src'] as $url ) {
+					// Add data URIs first.
+					if ( 0 === strpos( trim( $url ), 'data:' ) ) {
+						$src_ordered[] = array(
+							'url'    => $url,
+							'format' => 'data',
+						);
+						continue;
+					}
+					$format         = pathinfo( $url, PATHINFO_EXTENSION );
+					$src[ $format ] = $url;
+				}
+
+				// Add woff2.
+				if ( ! empty( $src['woff2'] ) ) {
+					$src_ordered[] = array(
+						'url'    => $src['woff2'],
```

Comment:
```suggestion
						'url'    => sanitize_url( $src['woff2'] ),
```

Possibly `esc_url()` but I think CSS and HTML escape URLs differently.

I think `file` protocol has been normalized to an actual URL at this point, they'll get stripped if not. data URIs can't be sanitized by core as far as I know.

:1234: Applies to this and the items below.

Reply:
Resolved https://github.com/WordPress/gutenberg/pull/40493/commits/4f6f098598f99a14a400920847f7dc2666c70043